src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/Streaming/DataStreamWriterTests.cs

```diff
@@ -71,44 +71,45 @@ public void TestSignaturesV2_3_X()
         [SkipIfSparkVersionIsLessThan(Versions.V2_4_0)]
         public void TestForeachBatch()
         {
-            // Temporary folder to put our test stream input.
-            using var srcTempDirectory = new TemporaryDirectory();
+            var memoryStream = new MemoryStream<int>(_spark);
+
             // Temporary folder to write ForeachBatch output.
             using var dstTempDirectory = new TemporaryDirectory();
 
             Func<Column, Column> outerUdf = Udf<int, int>(i => i + 100);
 
-            // id column: [0, 1, ..., 9]
-            WriteCsv(0, 10, Path.Combine(srcTempDirectory.Path, "input1.csv"));
-
-            DataStreamWriter dsw = _spark
-                .ReadStream()
-                .Schema("id INT")
-                .Csv(srcTempDirectory.Path)
+            DataStreamWriter dsw = memoryStream
+                .ToDF()
                 .WriteStream()
                 .ForeachBatch((df, id) =>
                 {
                     Func<Column, Column> innerUdf = Udf<int, int>(i => i + 200);
-                    df.Select(outerUdf(innerUdf(Col("id"))))
+                    df.Select(outerUdf(innerUdf(Col("value"))))
                         .Write()
                         .Csv(Path.Combine(dstTempDirectory.Path, id.ToString()));
                 });
 
             StreamingQuery sq = dsw.Start();
 
-            // Process until all available data in the source has been processed and committed
-            // to the ForeachBatch sink. 
-            sq.ProcessAllAvailable();
```

Comment:
Can you update the PR description why this didn't work?

Reply:
I don't know if this was what was failing the tests, but this is something that some spark tests do when using `ProcessAllAvailable`

Reply:
Oh then why not just put `FailAfter` with the current approach without `MemoryStream`?

Reply:
Spark tests also use `MemoryStream`.  If this test is flaky, then hoping this will fix it.

Reply:
Unless you are absolutely sure (you mentioned "some" spark tests were using this not all, so I am confused), can we first use `FailAfter` to monitor if this is causing the hang instead of brining in other changes?

Reply:
Also, do you think we can try to repro this locally? (e.g., run this in a loop, etc.)